File: code_examples/DesignPatterns/Prototype.js
function cocktailPrototype(name, ingredients) {
    function text() {
        console.log(`Cocktail: ${this.name}\nIngredients: ${this.ingredients.join(', ')})`)
    }
    return {
        name,
        ingredients,
        text,
    };
}

const frenchMartiniPrototype = cocktailPrototype('French Martini', ['Vodka', 'Chambord', 'Pineapple Juice', 'Licor 75']);
const darkAndStormyPrototype = cocktailPrototype('Dark and Stormy', ['Dark Rum', 'Ginger Beer', 'Lime juice']);
const whiskySourPrototype = cocktailPrototype('Whisky Sour', ['Bourbon', 'Lemon Juice', 'Sugar Syrup']);

const frenchMartini = Object.create(frenchMartiniPrototype);
frenchMartini.name = 'French Martini';
frenchMartini.ingredients.pop();

const darkAndStormy = Object.create(darkAndStormyPrototype);
darkAndStormy.name = 'Dark and Stormy';

const whiskySour = Object.create(whiskySourPrototype);
whiskySour.name = 'Whisky Sour';
whiskySour.ingredients.push('Egg White');

frenchMartini.text();
darkAndStormy.text();
whiskySour.text();

File: code_examples/DesignPatterns/prototype.kt
class CocktailPrototype(var name: String, var ingredients: MutableList<String>) {
    fun text() {
        println("Cocktail: $name \nIngredients: ${ingredients.joinToString(", ")}")
    }
}

fun main() {
    val frenchMartiniPrototype = CocktailPrototype("French Martini", mutableListOf("Vodka", "Chambord", "Pineapple Juice", "Licor 75"))
    val darkAndStormyPrototype = CocktailPrototype("Dark and Stormy", mutableListOf("Dark Rum", "Lemons", "Ginger Beer", "Bitters"))
    val whiskeySourPrototype = CocktailPrototype("Whiskey Sour", mutableListOf("Bourbon", "Egg White", "Sugar Syrup", "Lemon Juice"))

    val frenchMartini = CocktailPrototype(frenchMartiniPrototype.name, frenchMartiniPrototype.ingredients.toMutableList())
    frenchMartini.ingredients.remove("Licor 75")
    frenchMartini.text()

    val darkAndStormy = CocktailPrototype(darkAndStormyPrototype.name, darkAndStormyPrototype.ingredients.toMutableList())
    darkAndStormy.ingredients.add("Lime Wedge")
    darkAndStormy.ingredients.remove("Lemons")
    darkAndStormy.text()

    val whiskeySour = CocktailPrototype(whiskeySourPrototype.name, whiskeySourPrototype.ingredients.toMutableList())
    whiskeySour.text()
}
